smarts/core/utils/traffic_history_serivce.py

```diff
@@ -0,0 +1,153 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+from dataclasses import dataclass
+from multiprocessing import Pipe, Process, Queue
+
+import ijson
+
+import smarts.core.scenario as scenario
+
+
+@dataclass
+class RequestHistoryRange:
+    start_index: int
+    end_index: int
+
+
+class Traffic_history_service:
+    """responsible for dynamically fetching traffic history json to reduce
+    memory use of traffic history data
+    """
+
+    def __init__(self, history_file_path):
+        self._history_file_path = history_file_path
+        self._all_timesteps = set()
+        self._current_traffic_history = {}
+        self._prev_batch_history = {}
+        # return if traffic history is not used
+        if history_file_path is None:
+            return
+
+        send_data_conn, receive_data_conn = Pipe()
+        self._receive_data_conn = receive_data_conn
+        self._request_queue = Queue()
+        self._fetch_history_proc = Process(
+            target=self._fetch_history,
+            args=(
+                send_data_conn,
+                self._request_queue,
+                self._history_file_path,
+                len(self._all_timesteps),
+            ),
```

Comment:
This argument is not used:
```suggestion
                self._history_file_path,
            ),
```